# Conserved coexpression at single cell resolution across primate brains

Determining how gene coexpression networks change across species is an essential step toward identifying novel regulatory relationships that underlie phenotypic variability. We use single cell expression data from the middle temporal gyrus of five primates (human, chimp, gorilla, macaque and marmoset) to identify marker gene sets that robustly distinguish 57 homologous cell types, and calculate the similarity of gene coexpression networks across matched cell types.

While ortholog functional conservation is generally high, we find 24% of genes with extensive differences in expression patterns between human and non-human primates in one or more cell classes, which are also associated with multiple brain disorders. To validate these targeted cell-type observations within a population corpus, we performed co-expression meta-analysis across 19 species to track changes in gene expression unique to the human lineage. We observed that genes showing changed cellular profiles across species exhibit differential co-expression between species, reflecting the changed pattern of activity.

![web-server-fig-01](https://user-images.githubusercontent.com/46113011/190569276-f26f19cf-9d85-4757-b593-d57245ee53a6.png)

# Data
Raw data (snRNA-seq expression atlases of the middle temporal gyrus from five primates) can be downloaded from the NeMO archive (RRID:SCR_016152), and related pre-processing steps are outlined in the associated github repository (https://github.com/AllenInstitute/Great_Ape_MTG). This repository contains additional data files and scripts to recapitulate our analysis. Our datasets are also available for further exploration and analysis through an interactive web browser (https://gillisweb.cshl.edu/Primate_MTG_coexp/).
